### PR TITLE
Upgrade jest

### PR DESCRIPTION
**Pull request by sokra**
This seem to fix the hanging builds in jest

**What kind of change does this PR introduce?**
test, ci

**Did you add tests for your changes?**
not needed

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
nothing

### CODE REVIEW

**Comment by sokra** [comment]
@niieani why is node running reports this many many times:

```
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/index.js:200    - return new Promise((resolve, reject) => {
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:545 - .worker({
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:606 - const promise = this._processFile(hasteMap, map, mocks, filePath);
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:326 - .then(data => this._buildHasteMap(data))
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:550 - .then(
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:606 - const promise = this._processFile(hasteMap, map, mocks, filePath);
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:326 - .then(data => this._buildHasteMap(data))
```

Note that we `yarn link webpack` into ourselves and create a circular symlink this way. Could this be an issue?

**Comment by sokra** [comment]
```
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:197 - return new Promise(function(resolve, reject) {
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:30  - return _ref.apply(this, arguments);
/home/travis/build/webpack/webpack/node_modules/jest-cli/bin/jest.js:16         - require('../build/cli').run();
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:197 - return new Promise(function(resolve, reject) {
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:310 - return _ref3.apply(this, arguments);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:14  - var _ref2 = yield runCLI(argv, projects);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200 - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219 - return step('next');
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:197 - return new Promise(function(resolve, reject) {
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:629 - return _ref8.apply(this, arguments);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:261 - yield _run(
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200 - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219 - return step('next');
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:197 - return new Promise(function(resolve, reject) {
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:580 - return _ref6.apply(this, arguments);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:600 - var _ref9 = yield buildContextsAndHasteMaps(
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200 - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219 - return step('next');
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:197 - return new Promise(function(resolve, reject) {
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:570 - return _ref7.apply(this, arguments);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:545 - configs.map(
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200 - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219 - return step('next');
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:385 - return Promise.resolve()
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:325 - this._buildPromise = this._buildFileMap()
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:566   - .default)(config, yield hasteMapInstance.build());
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200   - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219   - return step('next');
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:386 - .then(() => read.call(this))
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:325 - this._buildPromise = this._buildFileMap()
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:566   - .default)(config, yield hasteMapInstance.build());
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200   - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219   - return step('next');
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:387 - .catch(() => this._createEmptyMap())
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:325 - this._buildPromise = this._buildFileMap()
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:566   - .default)(config, yield hasteMapInstance.build());
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200   - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219   - return step('next');
# PROMISE
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:388 - .then(cachedHasteMap => {
/home/travis/build/webpack/webpack/node_modules/jest-haste-map/build/index.js:325 - this._buildPromise = this._buildFileMap()
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:566   - .default)(config, yield hasteMapInstance.build());
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:200   - var info = gen[key](arg);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/cli/index.js:219   - return step('next');
```

**Comment by niieani** [comment]
> Note that we yarn link webpack into ourselves and create a circular symlink this way. Could this be an issue?

@sokra that seems like a possibility. I'm unfamiliar with Haste though.

Could anybody from the `jest` team confirm if this is a good lead? 
CC: @SimenB @cpojer @mjesun

**Comment by SimenB** [comment]
`why-is-node-running` will print a bunch of false positives as we have a bunch of promises here and there (every single test, within every single spec file). It would be interesting to see if `--detectLeaks` reports anything, which filters out Jest internals (it only tracks async operations from code executed in tests).

I'm not sure about the circular symlink, I'm not too familiar with the jest-haste-map. Might be, though!

**Comment by montogeek** [comment]
@SimenB There is only one `execModule` call in the entire log

**Comment by SimenB** [comment]
In theory, that's the only code required from within a test, then

**Comment by sokra** [comment]
I tried `--detectOpenHandles` but this didn't report anything (probably because the process didn't exit.

**Comment by montogeek** [comment]
It seems this is related with how Jest coverage works, it does not happen with a normal run.
Maybe coverage takes too much time?

**Comment by SimenB** [comment]
Might be that you need to exclude `node_modules` from coverage?

https://github.com/webpack/webpack/blob/800c46e89c914ead09554d773697dc28836a57f1/package.json#L161-L165

**Comment by montogeek** [comment]
Funny that --detectOpenHandles is based on why-is-node-running. 💃

**Comment by montogeek** [comment]
I got a successful run locally after fixing those prettier issues:
image

**Comment by sokra** [comment]
@montogeek It's working fine locally for me too. Problem is the CI build.

Here is a why-is-node-running output without all these Promises:

```
There are 6 handle(s) keeping the process running
# Timeout
/home/travis/build/webpack/webpack/setup/why-is-it-running.js:4 - setInterval(function() {
# Timeout
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/reporters/Status.js:98            - this._interval = setInterval(() => this._tick(), 1000);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/reporters/default_reporter.js:162 - this._status.runStarted(aggregatedResults, options);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/reporter_dispatcher.js:78         - reporter.onRunStart && (yield reporter.onRunStart(results, options));
# PROCESSWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80 - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:55 - this._initialize();
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/index.js:103 - const worker = new (_worker || _load_worker()).default(workerOptions);
/home/travis/build/webpack/webpack/node_modules/jest-runner/build/index.js:134 - const worker = new (_jestWorker || _load_jestWorker()).default(
# PIPEWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80 - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:55 - this._initialize();
# PROCESSWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80  - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:206 - this._initialize();
# PIPEWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80  - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:206 - this._initialize();
There are 1 handle(s) keeping the process running
# Timeout
/home/travis/build/webpack/webpack/setup/why-is-it-running.js:4 - setInterval(function() {
There are 1 handle(s) keeping the process running
# Timeout
/home/travis/build/webpack/webpack/setup/why-is-it-running.js:4 - setInterval(function() {
There are 6 handle(s) keeping the process running
# Timeout
/home/travis/build/webpack/webpack/setup/why-is-it-running.js:4 - setInterval(function() {
# Timeout
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/reporters/Status.js:98            - this._interval = setInterval(() => this._tick(), 1000);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/reporters/default_reporter.js:162 - this._status.runStarted(aggregatedResults, options);
/home/travis/build/webpack/webpack/node_modules/jest-cli/build/reporter_dispatcher.js:78         - reporter.onRunStart && (yield reporter.onRunStart(results, options));
# PROCESSWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80 - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:55 - this._initialize();
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/index.js:103 - const worker = new (_worker || _load_worker()).default(workerOptions);
/home/travis/build/webpack/webpack/node_modules/jest-runner/build/index.js:134 - const worker = new (_jestWorker || _load_jestWorker()).default(
# PIPEWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80 - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:55 - this._initialize();
# PROCESSWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80  - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:206 - this._initialize();
# PIPEWRAP
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:80  - const child = (_child_process || _load_child_process()).default.fork(
/home/travis/build/webpack/webpack/node_modules/jest-worker/build/worker.js:206 - this._initialize();
```

And the worker processes:

```
There are 1 handle(s) keeping the process running
# Timeout
/home/travis/build/webpack/webpack/setup/why-is-it-running.js:4 - setInterval(function() {
There are 1 handle(s) keeping the process running
# Timeout
/home/travis/build/webpack/webpack/setup/why-is-it-running.js:4 - setInterval(function() {

**Comment by sokra** [comment]
Ok I figured out that this is related to the `ConfigTestCases` test suite and only occurs when running in a thread pool

**Comment by montogeek** [comment]
What is different about that test suite? Maybe that timeout in `webpack` callback?

**Comment by sokra** [comment]
hmm... I tried running each test suite separately, and this worked fine...

**Comment by sokra** [comment]
Ok I narrowed it down with a high probablilty to `ConfigTestCases plugins`.

My wild guess is now that it somehow infers with the UglifyJsPlugins which uses a worker poll to run minimization in parallel... I'll verify that now...

**Comment by niieani** [comment]
Painful to watch... 😵

**Comment by TheLarkInn** [comment]
I'm restarting the job I know it may not help but its driving me crazy seeing one test fail every time. :sad:

**Comment by sokra** [comment]
My guess was wrong. It's not caused by the Worker pool in the uglifyjs-plugin...

**Comment by ooflorent** [comment]
I’ve also observed timeout failures on stat test cases but they seem to be gone the day uglify was disabled.

File: .travis.yml [hunk 1]
```diff
@@ -10,49 +10,60 @@ branches:
...
-      env: NO_WATCH_TESTS=1 JEST=--maxWorkers=2 JOB_PART=integration
-      stage: versions
+      node_js: "10"
+      env: NO_WATCH_TESTS=1 JEST=--maxWorkers=2 JOB_PART="ConfigTestCases.test.js|Stats.test.js -t plugins|output|Stats"
+  allow_failures:
+    - os: osx
   fast_finish: true
 
+before_install:
+  - curl -o- -L https://yarnpkg.com/install.sh | bash
```
Comment (ooflorent) [inline .travis.yml]: Why are you installing yarn? This caused failures in the past

File: .travis.yml [hunk 2]
```diff
@@ -10,49 +10,60 @@ branches:
 cache:
   yarn: true
 
-stages:
```
Comment (ooflorent) [inline .travis.yml]: What was wrong with the multi stage build?

Comment (sokra) [reply .travis.yml]: nothing. I just run travis build to find the problem.

File: .travis.yml (hunk 1 above)
Comment (sokra) [reply .travis.yml]: hmm.. I'll remove it, but the problem was there before this change...

**Comment by sokra** [comment]
It doesn't seem to be related to `--maxWorkers 2`. It also occurs with `--runInBand`.

```
$ yarn cover:single --ci $JEST $JOB_PART
yarn run v1.3.2
(node:3466) [DEP0005] DeprecationWarning: Buffer() is deprecated due to security and usability issues. Please use the Buffer.alloc(), Buffer.allocUnsafe(), or Buffer.from() methods instead.
$ node -r ./setup/why-is-it-running.js --max-old-space-size=4096 node_modules/jest-cli/bin/jest --coverage --ci --runInBand --verbose 'ConfigTestCases.test.js|Stats.test.js'
No output has been received in the last 10m0s, this potentially indicates a stalled build or something wrong with the build itself.
Check the details on how to adjust your build configuration on: https://docs.travis-ci.com/user/common-build-problems/#Build-times-out-because-no-output-was-received
The build has been terminated
```

**Comment by ooflorent** [comment]
Maybe there is a really issue in webpack's codebase. Some kind of race condition...

**Comment by sokra** [comment]
> Maybe there is a really issue in webpack's codebase. Some kind of race condition...

But shouldn't jest handle that regardless of any error in webpack?

How does jest behave if my test is like `while(true);`?

**Comment by ooflorent** [comment]
I think jest would hang forever. React manages to catch this using a Babel transformer but it only works for synchronous code.

**Comment by montogeek** [comment]
7 of 8 Forcing exiting, It shouldn't happen, right?

**Comment by SimenB** [comment]
@sokra so sorry you're having issues! 😢

If the issue is tests hanging forever, you could consider a custom node environment where you overwrite `runScript` to be add a timeout.

See https://github.com/facebook/jest/blob/15ed971dc41ca8f2611eaf41e498125d8d894b27/packages/jest-environment-node/src/index.js#L93-L99

Just extend the env, but set `return script.runInContext(this.context);` to be `return script.runInContext(this.context, {timeout: 5 * 60 * 1000});` or something. That makes the runtime throw a timeout, even for `while(true){}` scripts. Might help you track down which test is hanging, at least

**Comment by montogeek** [comment]
All ended successfully, without force exiting.

**Comment by ooflorent** [comment]
@montogeek This happens often. What's important here is consistency.

**Comment by sokra** [comment]
I wasn't able to fix the problem, but I worked around it by splitting the tests into two travis/appveyor jobs.

Seems like webpack compilation is busy hanging when compiling too often in the same process (maybe memory issues, leaks, i don't know).

It would be great to have a jest option to restart workers after each test suite. This should be possible to add.

**Comment by rickhanlonii** [comment]
@sokra does this work in a single job with `--runInBand`?

**Comment by rickhanlonii** [comment]
@sokra I ask because I've seen similar issues in CI when not passing the explicit number of workers available in the vm ([we'll use the number of CPUs](https://github.com/facebook/jest/blob/77744a24816d0978b6c478987426c36d615864bd/packages/jest-config/src/get_max_workers.js#L20) on the machine, which is more than the number of CPUs available to the VM)

Try running with `-i` to confirm or `-w=<CPUs given to the VM>` to fix

**Pull request closed without merging**